# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [ ]:
import pandas as pd
import plotly.express as px


# Notebook Presentation

In [ ]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [ ]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [ ]:
print(f'Size of the df is {df_apps.shape}')
df_apps.sample(5)

Size of the df is (10841, 12)


App Category  Rating  Reviews  Size_MBs  \
2978  U of I Community Credit Union  FINANCE    3.70       52     13.00   
9255   magicApp Calling & Messaging   SOCIAL    4.10   207712      5.30   
9476  PES 2018 PRO EVOLUTION SOCCER   SPORTS    4.40  1721943     26.00   
3290  BT Remote (SPP) for Bluetooth    TOOLS    4.10       36      3.10   
443                        In to cm    TOOLS     NaN        0      2.30   

        Installs  Type Price Content_Rating   Genres      Last_Updated  \
2978       5,000  Free     0       Everyone  Finance       May 4, 2018   
9255  10,000,000  Free     0       Everyone   Social      June 9, 2018   
9476  10,000,000  Free     0       Everyone   Sports     June 27, 2018   
3290      10,000  Free     0       Everyone    Tools  October 24, 2016   
443           10  Free     0       Everyone    Tools     April 5, 2018   

       Android_Ver  
2978    4.4 and up  
9255    2.3 and up  
9476    5.0 and up  
3290  2.3.3 and up  
443   4.0.3 and up

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns. 

In [ ]:
df_apps_clean = df_apps.drop(['Last_Updated','Android_Ver'], axis=1)
df_apps.head(5)

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0      8.70        0  Paid  $13.99           Teen           Social   
1        0     33.00        0  Paid   $2.99       Everyone        Education   
2        0      5.50        0  Paid   $1.49       Everyone  Personalization   
3        0     19.00        0   NaN       0   Everyone 10+         Strategy   
4        0     10.00        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows. 

In [ ]:
print(df_apps_clean.isna().values.sum())

1475


In [ ]:
df_apps_clean = df_apps_clean.dropna()
print(df_apps_clean.isna().values.sum())
df_apps_clean.head(5)

0


App Category  Rating  Reviews  \
21                              KBA-EZ Health Guide  MEDICAL    5.00        4   
28                                         Ra Ga Ba     GAME    5.00        2   
47                                          Mu.F.O.     GAME    5.00        2   
82                                 Brick Breaker BR     GAME    5.00        7   
99  Anatomy & Physiology Vocabulary Exam Review App  MEDICAL    5.00        1   

    Size_MBs Installs  Type  Price Content_Rating   Genres  
21     25.00        1  Free      0       Everyone  Medical  
28     20.00        1  Paid  $1.49       Everyone   Arcade  
47     16.00        1  Paid  $0.99       Everyone   Arcade  
82     19.00        5  Free      0       Everyone   Arcade  
99      4.60        5  Free      0       Everyone  Medical

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`. 


In [ ]:
print(df_apps_clean.duplicated().values.sum())

476


In [ ]:
df_apps_clean.shape

(9367, 10)

In [ ]:
df_apps = df_apps_clean
df_apps_clean = df_apps_clean.drop_duplicates(subset=['App', 'Type', 'Price'])
df_apps_clean.shape

(8199, 10)

In [ ]:
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [ ]:
df_high_rated = df_apps_clean.sort_values(['Rating','Reviews'], ascending=False)
df_high_rated.head(10)

App            Category  \
2095                                         Ríos de Fe           LIFESTYLE   
2438  FD Calculator (EMI, SIP, RD & Loan Eligilibility)             FINANCE   
3115                                         Oración CX           LIFESTYLE   
2107                     Barisal University App-BU Face              FAMILY   
2069                                         Master E.K              FAMILY   
1968                                            CL REPL               TOOLS   
790                                              AJ Cam         PHOTOGRAPHY   
1275   AI Today : Artificial Intelligence News & AI 101  NEWS_AND_MAGAZINES   
1789                                            Ek Vote        PRODUCTIVITY   
2544                        CS & IT Interview Questions              FAMILY   

      Rating  Reviews  Size_MBs Installs  Type Price Content_Rating  \
2095    5.00      141     15.00    1,000  Free     0       Everyone   
2438    5.00      104      2.30    1,000  Free     0       Everyone   
3115    5.00      103      3.80    5,000  Free     0       Everyone   
2107    5.00      100     10.00    1,000  Free     0       Everyone   
2069    5.00       90     19.00    1,000  Free     0       Everyone   
1968    5.00       47     17.00    1,000  Free     0       Everyone   
790     5.00       44      2.80      100  Free     0       Everyone   
1275    5.00       43      2.30      100  Free     0       Everyone   
1789    5.00       43      6.20      500  Free     0       Everyone   
2544    5.00       43      3.30    1,000  Free     0       Everyone   

                Genres  
2095         Lifestyle  
2438           Finance  
3115         Lifestyle  
2107         Education  
2069         Education  
1968             Tools  
790        Photography  
1275  News & Magazines  
1789      Productivity  
2544         Education

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please? 

In [ ]:
df_high_rated = df_apps_clean.sort_values(['Size_MBs'], ascending=False)
df_high_rated.head(5)

App            Category  Rating  Reviews  \
9942   Talking Babsy Baby: Baby Games           LIFESTYLE    4.00   140995   
10687          Hungry Shark Evolution                GAME    4.50  6074334   
9943            Miami crime simulator                GAME    4.00   254518   
9944     Gangster Town: Vice District              FAMILY    4.30    65146   
3144                       Vi Trainer  HEALTH_AND_FITNESS    3.60      124   

       Size_MBs     Installs  Type Price Content_Rating  \
9942     100.00   10,000,000  Free     0       Everyone   
10687    100.00  100,000,000  Free     0           Teen   
9943     100.00   10,000,000  Free     0     Mature 17+   
9944     100.00   10,000,000  Free     0     Mature 17+   
3144     100.00        5,000  Free     0       Everyone   

                       Genres  
9942   Lifestyle;Pretend Play  
10687                  Arcade  
9943                   Action  
9944               Simulation  
3144         Health & Fitness

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [ ]:
df_high_rated = df_apps_clean.sort_values(['Reviews'], ascending=False)
df_high_rated.head(5)

App       Category  Rating  \
10805                                  Facebook         SOCIAL    4.10   
10785                        WhatsApp Messenger  COMMUNICATION    4.40   
10806                                 Instagram         SOCIAL    4.50   
10784  Messenger – Text and Video Chat for Free  COMMUNICATION    4.00   
10650                            Clash of Clans           GAME    4.60   

        Reviews  Size_MBs       Installs  Type Price Content_Rating  \
10805  78158306      5.30  1,000,000,000  Free     0           Teen   
10785  69119316      3.50  1,000,000,000  Free     0       Everyone   
10806  66577313      5.30  1,000,000,000  Free     0           Teen   
10784  56642847      3.50  1,000,000,000  Free     0       Everyone   
10650  44891723     98.00    100,000,000  Free     0   Everyone 10+   

              Genres  
10805         Social  
10785  Communication  
10806         Social  
10784  Communication  
10650       Strategy

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [ ]:
ratings = df_apps_clean.Content_Rating.value_counts()
ratings



Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content_Rating, dtype: int64

In [ ]:
fig = px.pie(labels=ratings.index,
             values=ratings.values,
             title="Content Rating",
             names=ratings.index,
             hole=0.6
             )
fig.update_traces(textposition='outside', textinfo='percent+label')
fig.show()

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install? 

Check the datatype of the Installs column.

Count the number of apps at each level of installations. 

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first. 

In [ ]:
df_apps_clean = df_apps_clean.sort_values(['Installs'], ascending=False)
df_apps_clean.head(5)
df_apps_clean.Installs.describe()

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [ ]:
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs           
1                 3
1,000           698
1,000,000      1417
1,000,000,000    20
10               69
10,000          988
10,000,000      933
100             303
100,000        1096
100,000,000     189
5                 9
5,000           425
5,000,000       607
50               56
50,000          457
50,000,000      202
500             199
500,000         504
500,000,000      24

In [ ]:
df_apps_clean.Installs = df_apps_clean.Installs.astype(str).str.replace(',', "")
df_apps_clean.Installs = pd.to_numeric(df_apps_clean.Installs)
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [ ]:
df_apps_clean.Price.describe()

count     8199
unique      73
top          0
freq      7595
Name: Price, dtype: object

### The most expensive apps sub $250

In [ ]:
df_apps_clean.Price = df_apps_clean.Price.astype(str).str.replace('$', "")
df_apps_clean.Price = pd.to_numeric(df_apps_clean.Price)
 
df_apps_clean.sort_values('Price', ascending=False).head(20)

<ipython-input-21-9f696e53366f>:1: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.



App   Category  Rating  Reviews  Size_MBs  \
3946        I'm Rich - Trump Edition  LIFESTYLE    3.60      275      7.30   
4606               I Am Rich Premium    FINANCE    4.10     1867      4.70   
2461              I AM RICH PRO PLUS    FINANCE    4.00       36     41.00   
1331          most expensive app (H)     FAMILY    4.30        6      1.50   
1946  I am rich (Most expensive app)    FINANCE    4.10      129      2.70   
3221                  I am Rich Plus     FAMILY    4.00      856      8.70   
2394                      I am Rich!    FINANCE    3.80       93     22.00   
3554                      💎 I'm rich  LIFESTYLE    3.80      718     26.00   
3114                       I am Rich    FINANCE    4.30      180      3.80   
3145              I am rich(premium)    FINANCE    3.50      472      0.94   
2775                   I Am Rich Pro     FAMILY    4.40      201      2.70   
5765                       I am rich  LIFESTYLE    3.80     3547      1.80   
3897                       I Am Rich     FAMILY    3.60      217      4.90   
2193             I am extremely Rich  LIFESTYLE    2.90       41      2.90   
3856                   I am rich VIP  LIFESTYLE    3.80      411      2.60   
2281       Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407                    LTC AS Legal    MEDICAL    4.00        6      1.30   
2629                I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481         A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
2207                        EMT PASS    MEDICAL    3.40       51      2.40   

      Installs  Type  Price Content_Rating         Genres  
3946     10000  Paid 400.00       Everyone      Lifestyle  
4606     50000  Paid 399.99       Everyone        Finance  
2461      1000  Paid 399.99       Everyone        Finance  
1331       100  Paid 399.99       Everyone  Entertainment  
1946      1000  Paid 399.99           Teen        Finance  
3221     10000  Paid 399.99       Everyone  Entertainment  
2394      1000  Paid 399.99       Everyone        Finance  
3554     10000  Paid 399.99       Everyone      Lifestyle  
3114      5000  Paid 399.99       Everyone        Finance  
3145      5000  Paid 399.99       Everyone        Finance  
2775      5000  Paid 399.99       Everyone  Entertainment  
5765    100000  Paid 399.99       Everyone      Lifestyle  
3897     10000  Paid 389.99       Everyone  Entertainment  
2193      1000  Paid 379.99       Everyone      Lifestyle  
3856     10000  Paid 299.99       Everyone      Lifestyle  
2281      1000  Paid  79.99       Everyone        Medical  
1407       100  Paid  39.99       Everyone        Medical  
2629      1000  Paid  37.99       Everyone      Lifestyle  
2481      1000  Paid  33.99       Everyone        Medical  
2207      1000  Paid  29.99       Everyone        Medical

### Highest Grossing Paid Apps (ballpark estimate)

In [ ]:
df_apps_clean = df_apps_clean[df_apps_clean['Price'] < 250]
df_apps_clean.sort_values('Price', ascending=False).head(5)
df_apps_clean['Revenue_Estimate'] = df_apps_clean.Installs.mul(df_apps_clean.Price)
df_apps_clean.sort_values('Revenue_Estimate', ascending=False)[:10]

<ipython-input-22-81d750d3955e>:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



App       Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft         FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper           GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas           GAME    4.40   348962     26.00   
7477            Facetune - For Free    PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock      LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator           GAME    4.60    87766     12.00   
6082                   Weather Live        WEATHER    4.50    76593      4.75   
7633        Five Nights at Freddy's           GAME    4.60   100805     50.00   
7044                         Tasker          TOOLS    4.60    43045      3.40   
6856                        Threema  COMMUNICATION    4.50    51110      3.50   

      Installs  Type  Price Content_Rating                     Genres  \
9220  10000000  Paid   6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid   0.99     Mature 17+                     Action   
7151   1000000  Paid   6.99     Mature 17+                     Action   
7477   1000000  Paid   5.99       Everyone                Photography   
7977   1000000  Paid   5.99       Everyone                  Lifestyle   
6594   1000000  Paid   4.99       Everyone                     Action   
6082    500000  Paid   5.99       Everyone                    Weather   
7633   1000000  Paid   2.99           Teen                     Action   
7044   1000000  Paid   2.99       Everyone                      Tools   
6856   1000000  Paid   2.99       Everyone              Communication   

      Revenue_Estimate  
9220     69,900,000.00  
8825      9,900,000.00  
7151      6,990,000.00  
7477      5,990,000.00  
7977      5,990,000.00  
6594      4,990,000.00  
6082      2,995,000.00  
7633      2,990,000.00  
7044      2,990,000.00  
6856      2,990,000.00

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

In [ ]:
df_apps_clean.Category.nunique()

33

In [ ]:
top10_category = df_apps_clean.Category.value_counts()[:10]
top10_category

FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: Category, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [ ]:
bar = px.bar(x = top10_category.index, # index = category name
             y = top10_category.values)
 
bar.show()

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [ ]:
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

In [ ]:
h_bar = px.bar(x = category_installs.Installs,
               y = category_installs.index,
               orientation='h',
               title='Category Popularity')
 
h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')
h_bar.show()

### Category Concentration - Downloads vs. Competition

**Challenge**: 
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this. 

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log') 

In [ ]:
cat_number = df_apps_clean.groupby('Category').agg({'App': pd.Series.count})
cat_merged_df = pd.merge(cat_number, category_installs, on='Category', how="inner")
print(f'The dimensions of the DataFrame are: {cat_merged_df.shape}')
cat_merged_df.sort_values('Installs', ascending=False)
scatter = px.scatter(cat_merged_df, # data
                    x='App', # column name
                    y='Installs',
                    title='Category Concentration',
                    size='App',
                    hover_name=cat_merged_df.index,
                    color='Installs')
 
scatter.update_layout(xaxis_title="Number of Apps (Lower=More Concentrated)",
                      yaxis_title="Installs",
                      yaxis=dict(type='log'))
 
scatter.show()

The dimensions of the DataFrame are: (33, 2)


# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html). 


In [ ]:
len(df_apps_clean.Genres.unique())

114

In [ ]:
df_apps_clean.head(5)

App       Category  \
10731                                     My Talking Tom           GAME   
10746  Security Master - Antivirus, VPN, AppLock, Boo...          TOOLS   
10711                         SHAREit - Transfer & Share          TOOLS   
10713                      imo free video calls and chat  COMMUNICATION   
10717                                                Pou           GAME   

       Rating   Reviews  Size_MBs   Installs  Type  Price Content_Rating  \
10731    4.50  14891223     36.00  500000000  Free   0.00       Everyone   
10746    4.70  24900999      3.40  500000000  Free   0.00       Everyone   
10711    4.60   7790693     17.00  500000000  Free   0.00       Everyone   
10713    4.30   4785892     11.00  500000000  Free   0.00       Everyone   
10717    4.30  10485308     24.00  500000000  Free   0.00       Everyone   

              Genres  Revenue_Estimate  
10731         Casual              0.00  
10746          Tools              0.00  
10711          Tools              0.00  
10713  Communication              0.00  
10717         Casual              0.00

If we look at the number of unique values in the Genres column we get 114. But this is not accurate if we have nested data like we do here. We can see this using .value_counts() and looking at the values that just have a single entry. There we see that the semi-colon (;) separates the genre names.

In [ ]:
df_apps_clean.Genres.value_counts().sort_values(ascending=True)[:5]

Racing;Pretend Play                    1
Parenting;Brain Games                  1
Entertainment;Education                1
Strategy;Creativity                    1
Health & Fitness;Action & Adventure    1
Name: Genres, dtype: int64

We somehow need to separate the genre names to get a clear picture. This is where the string’s .split() method comes in handy. After we’ve separated our genre names based on the semi-colon, we can add them all into a single column with .stack() and then use .value_counts().

In [ ]:
# Split the strings on the semi-colon and then .stack them.
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
print(f'We now have a single column with shape: {stack.shape}')
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')

We now have a single column with shape: (8564,)
Number of genres: 53


# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data? 

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/). 

* Find a way to set the colour scale using the color_continuous_scale parameter. 
* Find a way to make the color axis disappear by using coloraxis_showscale. 

In [ ]:
print(len(num_genres))
print(stack)
num_genres.head(5)

53
10731  0             Casual
10746  0              Tools
10711  0              Tools
10713  0      Communication
10717  0             Casual
                 ...       
2403   0    Personalization
2402   0              Tools
28     0             Arcade
47     0             Arcade
21     0            Medical
Length: 8564, dtype: object


Tools            719
Education        587
Entertainment    498
Action           304
Productivity     301
dtype: int64

In [ ]:
bar = px.bar(x = num_genres.index[:15], # index = category name
             y = num_genres[:15].values,
             title = 'Top Genres',
             hover_name = num_genres.index[:15],
             color = num_genres.values[:15],
             color_continuous_scale = 'cividis'
             )
bar.update_layout(xaxis_title='Genre',
                  yaxis_title='Number of Apps',
                  coloraxis_showscale=False)
 
bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [ ]:
df_apps_clean.Type.value_counts()

Free    7595
Paid     589
Name: Type, dtype: int64

In [ ]:
df_free_vs_paid_apps = df_apps_clean.groupby(["Category", "Type"], as_index=False).agg({'App': pd.Series.count})
df_free_vs_paid_apps.head()

Category  Type  App
0     ART_AND_DESIGN  Free   58
1     ART_AND_DESIGN  Paid    3
2  AUTO_AND_VEHICLES  Free   72
3  AUTO_AND_VEHICLES  Paid    1
4             BEAUTY  Free   42

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart: 

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category. 

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value). 

In [ ]:
g_bar = px.bar(df_free_vs_paid_apps,
               x='Category',
               y='App',
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group')
 
g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder':'total descending'},
                    yaxis=dict(type='log'))
 
g_bar.show()

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart. 

<img src=https://imgur.com/uVsECT3.png>


In [ ]:
box = px.box(df_apps_clean,
             y='Installs',
             x='Type',
             color='Type',
             notched=True,
             points='all',
             title='How Many Downloads are Paid Apps Giving Up?')
 
box.update_layout(yaxis=dict(type='log'))
 
box.show()

# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below: 

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories. 

In [ ]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
box = px.box(df_paid_apps, 
             x='Category', 
             y='Revenue_Estimate',
             title='How Much Can Paid Apps Earn?')
 
box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))
 
 
box.show()

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [ ]:
box = px.box(df_paid_apps,
             x='Category',
             y="Price",
             title='Price per Category')
 
box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Price',
                  xaxis={'categoryorder':'max descending'},
                  yaxis=dict(type='log'))
 
box.show()

<h2>Learning Points & Summary</h2>
<h3>In this lesson we looked at how to:</h3>
<ul>
<li>
Pull a random sample from a DataFrame using .sample()
</li>
<li>
How to find duplicate entries with .duplicated() and .drop_duplicates()
</li>
<li>
How to convert string and object data types into numbers with .to_numeric()
</li>
<li>
How to use plotly to generate beautiful pie, donut, and bar charts as well as box and scatter plots
</li>
</ul>

Well done on completing today's lessons! 👏👏👏 I hope you enjoyed today’s Google Play Store analysis. It's incredible how much more of a clear picture we can get about the app ecosystem just by looking into some scraped website data for the Google Play Store.

